### DISCUSSION TITLE
Expose runner version in the runner context

### DISCUSSION THREAD

**Discussion by scottohara** [community/community#26440] [Actions]
<p><strong>Request:</strong></p>
<p>In the <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#runner-context" rel="nofollow noopener">runner context object</a>, add <code>version</code> to allow jobs to access the version of <a href="https://github.com/actions/runner" rel="nofollow noopener">actions/runner</a> that the job currently being run with.</p>
<p><strong>Use case:</strong></p>
<p>I have a job that runs on macOS-latest and produces an artifact containing code coverage results (<code>lcov.info</code>). The coverage payload contains absolute file paths.</p>
<p>For example, in a GitHub repo named <code>foo</code> where one of the source files is <code>src/bar.ts</code>, the LCOV report would include something like:</p>
<pre><code>SF:/Users/runner/runners/2.164.0/work/foo/foo/src/bar.ts
</code></pre>
<p>Later in the workflow, another job (running on ubuntu-latest) collects the lcov.info file(s) produced from earlier jobs, formats &amp; merges them together into a single coverage report, and uploads them them to my code analysis service of choice (CodeClimate).</p>
<p>The issue is that <a href="https://github.com/codeclimate/test-reporter/blob/master/man/cc-test-reporter-format-coverage.1.md" rel="nofollow noopener">CodeClimate’s CLI utility</a> for formatting raw LCOV files needs to strip the prefix from absolute paths in the coverage payload, making them relative to the project root.</p>
<p>The command provides a <code>--prefix</code> argument that can be used to specify the prefix to strip.</p>
<p>An example of what the format command currently looks like is:</p>
<pre><code>jobs:
job1:
runs-on: macOS-latest
steps:
...

job2:
    runs-on: ubuntu-latest
    needs: job1
    steps:
      - run: ./cc-test-reporter format-coverage --prefix /Users/runner/runners/2.164.0/work/foo/foo --input-type lcov --output coverage.json lcov.info
</code></pre>
<p>The problem is the runner version (2.164.0) hardcoded into the above command.  Every few weeks the workflow YAML needs to be updated when a new actions/runner version is released.</p>
<p>Ideally, I’d like to be able to do something like:</p>
<pre><code>jobs:
  job1:
    runs-on: macOS-latest
    steps:
      ...

  job2:
    runs-on: ubuntu-latest
    needs: job1
    steps:
      - run: ./cc-test-reporter format-coverage --prefix /Users/runner/runners/${{ runner.version }}/work/foo/foo --input-type lcov --output coverage.json lcov.info
</code></pre>
<p>…replacing the hardcoded version number with a ${{ runner.version }} expression.</p>
<p>This assumes that the runner version for all virtual environments is the same (so a job running on macOS-latest and another job running on ubuntu-latest would share the same runner version). I’m not sure if that is the case or not.</p>
<p>Unfortunately, the runner context doesn’t expose the current version number as far as I can tell.</p>
<p><strong>Workaround:</strong><br>
An alternative may be to add an output in the macOS job to save the current working directory; and reference that output later:</p>
<pre><code>jobs:
  job1:
    runs-on: macOS-latest
    steps:
      - run: echo "::set-output name=prefix::$PWD"

  job2:
    runs-on: ubuntu-latest
    needs: job1
    steps:
      - run: ./cc-test-reporter format-coverage --prefix ${{ needs.job1.outputs.prefix }} --input-type lcov --output coverage.json lcov.info
</code></pre>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/scottohara">@scottohara</a> ,</p>
<p>The directory " <strong>/Users/runner/runners/2.164.0/work/foo/foo</strong>" actually is the <strong>workspace</strong> (the default working directory, <strong>github.workspace</strong> ) for the <strong>job1</strong> that runs on <strong>macos-latest</strong> in your workflow, and you can easily use the property " <strong>github.workspace</strong>" of the <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener"><strong>github context</strong></a> to access this directory.</p>
<p>So, on <strong>job2</strong> , you do not need to set this directory as a hardcoded in the command line, as you mentioned you can use the <strong>outputs</strong>.</p>
<p>You can do as the following steps:<br>
<strong>1)</strong> On <strong>job1</strong> , use the syntax “<a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjobs_idoutputs" rel="nofollow noopener"><strong>jobs.&lt;jobs_id&gt;.outputs</strong></a>” to set a job-level output to store the <strong>workspace</strong> of <strong>job1</strong>.</p>
<p><strong>2)</strong> On <strong>job2</strong> , because <strong>job2</strong> needs <strong>job1</strong> must complete successfully before <strong>job2</strong> will run, you should use the syntax “<a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds" rel="nofollow noopener"><strong>jobs.&lt;job_id&gt;.needs</strong></a>” on <strong>job2</strong>.</p>
<p><strong>3)</strong> In the steps of <strong>job2</strong> , you can use property " <strong>needs.&lt;job id&gt;.outputs.&lt;output name&gt;</strong>" of the <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#needs-context" rel="nofollow noopener"><strong>needs context</strong></a> to access the job-level output set on <strong>job1</strong>.</p>
<p>A simple demo:</p>
<pre><code>jobs:
  job1:
    name: Job 1
    outputs:
      prefix: ${{ github.workspace }}
    runs-on: macos-latest
    steps:
     . . .

  job2:
    name: Job 2
    needs: [job1]
    runs-on: ubuntu-latest
    steps:
      - name: view the output of job1
        run: echo "needs.job1.outputs.prefix = ${{ needs.job1.outputs.prefix }}"
</code></pre>
<p><strong>NOTE:</strong> if you set the output in a step of  <strong>job1</strong> , you can’t directly access this output on <strong>job2</strong> , you still need to set a job-level output on <strong>job1</strong> , and pass the step-level output to the job-level output.</p>
<pre><code>jobs:
  job1:
    name: Job 1
    outputs:
      prefix: ${{ steps.set_output.outputs.dir_prefix }}
    runs-on: macos-latest
    steps:
      - name: set output
id: set_output
run: echo "::set-output name=dir_prefix::${{ github.workspace }}"

  job2:
    name: Job 2
    needs: [job1]
    runs-on: ubuntu-latest
    steps:
      - name: view the output of job1
        run: echo "needs.job1.outputs.prefix = ${{ needs.job1.outputs.prefix }}"
</code></pre>

**Comment by scottohara** [comment]
<p>Thanks so much <a class="mention" href="/brightran">@brightran</a> !</p>
<p>I’ll give this a try.</p>

**Comment by scottohara** [comment]
<p>I may have been too quick to mark this as the correct solution.</p>
<p>Unfortunately, when attempting to use ${{ github.workspace }} as a job output, the following warning appears at the end of job1:</p>
<pre><code>Complete job
  Evaluate and set job outputs
  ##[warning]Skip output 'workspace' since it may contain secret.
  Cleaning up orphan processes
</code></pre>
<p>Later in job2, the output is not available (empty).</p>
<p>According to the <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjobs_idoutputs" rel="nofollow noopener">documentation</a> (emphasis mine):</p>
<blockquote>
<p>Job outputs are strings, and job outputs containing expressions are evaluated on the runner at the end of each job. <strong>Outputs containing secrets are redacted on the runner and not sent to GitHub Actions</strong>.</p>
</blockquote>
<p>I’m assuming that the ‘secret’ is the name of the repo that appears in the workspace path, as I can see in other logs that this is masked out:</p>
<pre><code>Uploading artifact 'lcov.info' from '/Users/runner/runners/2.169.1/work/ ***/*** /coverage/frontend/lcov.info' for run #43
</code></pre>
<p>Any other suggestions?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/scottohara">@scottohara</a> ,</p>
<p>No, by default the <strong>workspace</strong> directory and the <strong>repository name</strong> will not be set as <strong>secrets</strong> , they are public and you can explicitly view their values from the <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener"><strong>github context</strong></a>.<br>
I just tested again minutes ago, I did not get any warning or error when I passed the <strong>${{ github.workspace }}</strong> to the job-level output and step-level output. And I also can fetch the correct value of the outputs in the subsequent job.<br>
I tested this on my public repository and private repository, all of them can work fine as expected.<br>
The example I tested on my side: <a href="https://github.com/BrightRan/TestOutputs/actions/runs/92016491/workflow" rel="nofollow noopener">https://github.com/BrightRan/TestOutputs/actions/runs/92016491/workflow</a></p>
<p>You can try to check if you have set some secrets in your repository, and the value of a secret contains or equals the  <strong>repository name</strong>.</p>

**Comment by scottohara** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a>,</p>
<p>You are absolutely right…I <em>do</em> have a secret whose value matches the name of the repository, so that’s exactly what looks to be happening.</p>
<p>Luckily, I should be able to change the secret value to something else (it was only the same as the repo name for convenience); so I’ll give this a try.</p>
<p>(I’ll wait until I’ve confirmed this works before marking your earlier response as the solution, just to be sure :slight_smile:</p>
<p>Thanks again.</p>
